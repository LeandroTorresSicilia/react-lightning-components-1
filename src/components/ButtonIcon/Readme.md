ButtonIcon examples:

    <div>
        <div className="slds-p-bottom--medium">
            <ButtonIcon iconName="utility:like" onClick={ () => alert('clicked') } size="x-small" />
        </div>
        <div className="slds-p-bottom--medium">
            <ButtonIcon iconName="utility:settings" variant="dropdownInverse" title="setting dropdown button" style={{ backgroundColor: 'blue' }} />
        </div>
        <div className="slds-p-bottom--medium">
            <ButtonIcon iconName="utility:settings" variant="hinthover" assistiveText="approval button" />
        </div>
        <div className="slds-p-bottom--medium">
            <ButtonIcon iconName="custom:custom1" variant="border" />
        </div>
        <div className="slds-p-bottom--medium">
            <ButtonIcon iconName="standard:announcement" variant="borderInverse" style={{ backgroundColor: 'green' }} />
        </div>
    </div>