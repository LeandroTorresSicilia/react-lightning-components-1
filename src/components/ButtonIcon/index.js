/* eslint-disable react/require-default-props */
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import Button from '../Button';
import IconSvg from '../IconSvg';

const ICON_SVG_COMPONENT = (
    <IconSvg className="slds-button__icon slds-button__icon_x-small"
             iconName="utility:down"
             variant="bare" />
);

export default class ButtonIcon extends Component {
    getBtnClass() {
        const { className, size, variant } = this.props;

        return classnames('slds-button_icon', {
            'slds-button_icon-more': variant === 'dropdownBorder',
            'slds-button_icon-container-more': variant === 'dropdown',
            'slds-button_icon-inverse slds-button_icon-container-more': variant === 'dropdownInverse',
            'slds-button_icon-border': variant === 'border' || variant === 'hinthoverBorder',
            'slds-button_icon-border-inverse': variant === 'borderInverse' || variant === 'hinthoverInverse',
        }, {
            'slds-button_icon-small': size === 'small',
            'slds-button_icon-x-small': size === 'x-small',
            'slds-button_icon-xx-small': size === 'xx-small',
        }, className);
    }

    getIconSvgClass() {
        const { variant } = this.props;

        return classnames('slds-button__icon', {
            'slds-button__icon_hint': variant === 'hinthover' || variant === 'hinthoverBorder',
            'slds-button__icon_inverse-hint': variant === 'hinthoverInverse',
        });
    }

    renderComponent() {
        const {
            assistiveText,
            disabled,
            iconName,
            onClick,
            onFocus,
            onBlur,
            style,
            title,
            variant,
        } = this.props;
        const hasDropdownIcon = variant === 'dropdown' || variant === 'dropdownBorder' || variant === 'dropdownInverse';

        return (
            <Button {...this.props}
                    className={this.getBtnClass()}
                    ariaHaspopup={hasDropdownIcon}
                    disabled={disabled}
                    onClick={onClick}
                    onFocus={onFocus}
                    onBlur={onBlur}
                    style={style}
                    title={title}
                    variant="bare" >
                <IconSvg className={this.getIconSvgClass()} iconName={iconName} variant="bare" />
                { hasDropdownIcon ? ICON_SVG_COMPONENT : null }
                <span className="slds-assistive-text">{ assistiveText }</span>
            </Button>
        );
    }

    render() {
        const {
            size,
            variant,
        } = this.props;

        switch (variant) {
        case 'default':
        case 'border':
        case 'borderInverse':
            return this.renderComponent();
        case 'dropdown':
        case 'dropdownBorder':
        case 'dropdownInverse':
        /* This is because if a size prop if passed with another value,
         the dropdown button icon become a mess */
            if (size !== 'medium') {
                console.warn('size prop does not make sense for dropdown variant');
                return null;
            }
            return this.renderComponent();
        case 'hinthover':
        case 'hinthoverBorder':
        case 'hinthoverInverse':
            return (
                <div className="slds-hint-parent">
                    { this.renderComponent() }
                </div>
            );
        default:
            this.renderComponent();
        }
        return null;
    }
}

ButtonIcon.propTypes = {
    /** Description for assistive screen readers if is required */
    assistiveText: PropTypes.string,
    /** Class for custom styles */
    className: PropTypes.string,
    /** If is set to true it disables the button icon */
    disabled: PropTypes.bool,
    /** The icon name. It is required. It take the following format:
     ‘sprite name:icon name’ e.g. ‘utility:add’ */
    iconName: PropTypes.string.isRequired,
    /** Callback function fired when the button icon is clicked */
    onClick: PropTypes.func,
    /** Callback function fired when the button icon is focused */
    onFocus: PropTypes.func,
    /** Callback function fired when the button icon is blurred */
    onBlur: PropTypes.func,
    /** The button icon size. It don't make sense to dropdown variant */
    size: PropTypes.oneOf(['medium', 'small', 'x-small', 'xx-small']),
    /** Object with the custom styles. The properties must be in camelCase
     naming convention (e.g. { backgroundColor: green }) */
    style: PropTypes.object,
    /** The title that is showed when a user hover the icon. It is necessary
     if the button icon sits alone and a sighted user might need to hover for a description */
    title: PropTypes.string,
    /** This is the type of button. All inverse variants needs a background color style. */
    variant: PropTypes.oneOf([
        'default',
        'dropdown',
        'dropdownBorder',
        'dropdownInverse',
        'hinthover',
        'hinthoverBorder',
        'hinthoverInverse',
        'border',
        'borderInverse',
    ]),
};

ButtonIcon.defaultProps = {
    disabled: false,
    size: 'medium',
    variant: 'default',
    onClick: () => {},
    onFocus: () => {},
    onBlur: () => {},
};
