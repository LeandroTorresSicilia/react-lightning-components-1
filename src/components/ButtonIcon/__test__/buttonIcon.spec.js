import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import ButtonIcon from './../index';

describe('<ButtonIcon/>', () => {
    it('should render a button icon', () => {
        const component = renderer.create(<ButtonIcon iconName="utility:settings" />);

        expect(component).toMatchSnapshot();
    });

    it('should render a dropdown button icon', () => {
        const component = renderer.create(
            <ButtonIcon iconName="utility:settings" variant="dropdown" />,
        );

        expect(component).toMatchSnapshot();
    });

    it('should render a hinthover button icon', () => {
        const component = renderer.create(
            <ButtonIcon iconName="utility:settings" variant="hinthover" />,
        );

        expect(component).toMatchSnapshot();
    });

    it('should render a disabled button icon', () => {
        const component = renderer.create(
            <ButtonIcon iconName="utility:settings" disabled />);

        expect(component).toMatchSnapshot();
    });

    it('should render nothing when variant is one of the dropdown and size prop is passed other than medium', () => {
        const component = renderer.create(
            <ButtonIcon iconName="utility:settings" variant="dropdown" size="small" />,
        );

        expect(component).toMatchSnapshot();
    });

    it('should show a console warning when variant is one of the dropdown and size prop is passed other than medium', () => {
        console.warn = jest.fn();
        shallow(<ButtonIcon iconName="custom-custom5" variant="dropdownBorder" size="x-small" />);

        expect(console.warn).toHaveBeenCalledTimes(1);
    });

    it('should call onClick function when someone click over', () => {
        const onClickMockFn = jest.fn();
        const component = shallow(<ButtonIcon iconName="utility:like" onClick={onClickMockFn} />);

        component.simulate('click');
        expect(onClickMockFn.mock.calls.length).toBe(1);
    });

    it('should call onBlur function when it lost the focus', () => {
        const onBlurMockFn = jest.fn();
        const component = shallow(<ButtonIcon iconName="utility:like" onBlur={onBlurMockFn} />);

        component.simulate('blur');
        expect(onBlurMockFn.mock.calls.length).toBe(1);
    });

    it('should call onFocus function when it gets the focus', () => {
        const onFocusMockFn = jest.fn();
        const component = shallow(<ButtonIcon iconName="utility:like" onFocus={onFocusMockFn} />);

        component.simulate('focus');
        expect(onFocusMockFn.mock.calls.length).toBe(1);
    });
});
