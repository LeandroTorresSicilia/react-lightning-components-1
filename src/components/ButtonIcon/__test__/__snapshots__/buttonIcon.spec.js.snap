// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ButtonIcon/> should render a button icon 1`] = `
<button
  aria-haspopup={false}
  className="slds-button slds-button_icon"
  disabled={false}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  style={undefined}
  tabIndex={undefined}
  title={undefined}
>
  <svg
    aria-hidden="true"
    className="slds-button__icon"
    style={undefined}
  >
    <use
      xlinkHref="/icons/utility-sprite/svg/symbols.svg#settings"
      xmlnsXlink="http://www.w3.org/1999/xlink"
    />
  </svg>
  <span
    className="slds-assistive-text"
  />
</button>
`;

exports[`<ButtonIcon/> should render a disabled button icon 1`] = `
<button
  aria-haspopup={false}
  className="slds-button slds-button_icon"
  disabled={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  style={undefined}
  tabIndex={undefined}
  title={undefined}
>
  <svg
    aria-hidden="true"
    className="slds-button__icon"
    style={undefined}
  >
    <use
      xlinkHref="/icons/utility-sprite/svg/symbols.svg#settings"
      xmlnsXlink="http://www.w3.org/1999/xlink"
    />
  </svg>
  <span
    className="slds-assistive-text"
  />
</button>
`;

exports[`<ButtonIcon/> should render a dropdown button icon 1`] = `
<button
  aria-haspopup={true}
  className="slds-button slds-button_icon slds-button_icon-container-more"
  disabled={false}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  style={undefined}
  tabIndex={undefined}
  title={undefined}
>
  <svg
    aria-hidden="true"
    className="slds-button__icon"
    style={undefined}
  >
    <use
      xlinkHref="/icons/utility-sprite/svg/symbols.svg#settings"
      xmlnsXlink="http://www.w3.org/1999/xlink"
    />
  </svg>
  <svg
    aria-hidden="true"
    className="slds-button__icon slds-button__icon_x-small"
    style={undefined}
  >
    <use
      xlinkHref="/icons/utility-sprite/svg/symbols.svg#down"
      xmlnsXlink="http://www.w3.org/1999/xlink"
    />
  </svg>
  <span
    className="slds-assistive-text"
  />
</button>
`;

exports[`<ButtonIcon/> should render a hinthover button icon 1`] = `
<div
  className="slds-hint-parent"
>
  <button
    aria-haspopup={false}
    className="slds-button slds-button_icon"
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    style={undefined}
    tabIndex={undefined}
    title={undefined}
  >
    <svg
      aria-hidden="true"
      className="slds-button__icon slds-button__icon_hint"
      style={undefined}
    >
      <use
        xlinkHref="/icons/utility-sprite/svg/symbols.svg#settings"
        xmlnsXlink="http://www.w3.org/1999/xlink"
      />
    </svg>
    <span
      className="slds-assistive-text"
    />
  </button>
</div>
`;

exports[`<ButtonIcon/> should render nothing when variant is one of the dropdown and size prop is passed other than medium 1`] = `null`;
